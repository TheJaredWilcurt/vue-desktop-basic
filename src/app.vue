<template>
  <div id="app">

    <!-- example of getting data from the Vuex State -->
    <h1>{{ greeting }} World</h1>

    <!-- Vue Router  -->
    <nav>
      <router-link
        v-for="(route, index) in router.options.routes"
        v-if="route.component"
        :to="route.path"
        :key="index"
      >
        {{ route.name }}
      </router-link>
      |
      <a href="#" @click.prevent="openDevTools">Open Dev Tools</a>
    </nav>

    <router-view />

  </div>
</template>

<script>
module.exports = {
  name: 'app',
  data: function () {
    return {};
  },
  methods: {
    openDevTools: function () {
      nw.Window.get().showDevTools();
    }
  },
  computed: {
    greeting: function () {
      return this.$store.state.greeting;
    }
  }
};
</script>

<style>
.router-link-active {
  border: 1px solid #00F;
  border-radius: 50px;
  padding: 1px 9px 3px;
}
</style>
