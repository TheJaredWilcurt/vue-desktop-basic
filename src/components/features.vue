<template>
  <section>
    <h2>Features:</h2>
    <p><em>vue-desktop-basic</em> comes with:</p>
    <ul>
      <li v-for="feature in features">
        <a href="#" @click.prevent="openInDefaultBrowser(feature.url)">{{ feature.name }}</a>
      </li>
    </ul>
  </section>
</template>

<script>
  module.exports = {
    name: 'features',
    data: function () {
      return {
        features: [
          {
            name: 'NW.js (v' + nw.process.versions.nw + ')',
            url: 'https://nwjs.io'
          },
          {
            name: 'NW.js Builder (Phoenix)',
            url: 'https://github.com/evshiron/nwjs-builder-phoenix'
          },
          {
            name: 'Vue.js',
            url: 'https://vuejs.org'
          },
          {
            name: 'Vue-Router',
            url: 'https://router.vuejs.org/en/'
          },
          {
            name: 'Vuex',
            url: 'https://vuex.vuejs.org/en'
          },
          {
            name: 'HTTP-Vue-Loader',
            url: 'https://github.com/FranckFreiburger/http-vue-loader'
          }
        ]
      };
    },
    methods: {
      // Since your app is running in Chromium,
      // links to websites will navigate away from your app,
      // so you need to open them in the user's default browser
      openInDefaultBrowser: function (url) {
        nw.Shell.openExternal(url);
      }
    }
  };
</script>

