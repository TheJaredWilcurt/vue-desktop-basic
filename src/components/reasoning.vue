<template>
  <section>
    <h2>Reasoning:</h2>
    <p><em>Why not use a bundler, like WebPack/Browserify/Parcel/StealJS/Fuse-Box/Etc?</em></p>
    <ul>
      <li>Because bundlers are designed around the limitations of the web. We are working with a different medium, Cross-Platform Desktop Apps (XPDAs).</li>
      <li>The purpose of a bundler is to bundle our code. Features like concatenating, tree-shaking, uglifying, and minifying our code doesn't really offer any benefit. Nor does reducing the number of simultaneous network requests for assets. All assets are local, because desktop apps should follow the best practice of "offline-first".</li>
      <li>Ultimately you are shipping an 80+ MB app to your end user, so the tiny optimizations made for the medium of web are quite arbitrary here.</li>
      <li>A much greater concern for desktop apps would be the amount of dependencies in your <code>node_modules</code> folder, since all dependencies, and their dependencies are shipped to the end user (this excludes devDependencies). For more detailed information on optimizing for desktop apps read this article:
        <ul>
          <li><a href="#" @click.prevent="nw.Shell.openExternal('https://dev.to/thejaredwilcurt/reducing-app-distribution-size-in-nwjs-3d5f')">Reducing app distribution size in NW.js</a></li>
        </ul>
      </li>
      <li>If you wish to use additional features for development (linting, HMR, Sass, etc), I would recommend adding those features via a task runner, like Gulp, or just via simple npm scripts.</li>
    </ul>
    <p><em>Why not use a CDN for Vue/Vue-Router/Vuex/Etc instead of adding them as dependencies?</em></p>
    <ul>
      <li>Because for desktop apps you should always be designing with an "offline-first" mentality. Your app should still run if the user is in Airplane mode.</li>
      <li>Keeping them as dependencies allows you to easily handle version management.</li>
      <li>An alternative option would be to create a script that downloads the desired versions of these libraries into an untracked folder. This would mean you would have the desired versions, and no additional files or dependencies. Then run that script as a <code>postinstall</code> npm script. The downside to this is that it would be less familiar to other developers. Though technically more efficient, it is also slightly less maintable and requires documentation for future maintainers/contributors.</li>
    </ul>
  </section>
</template>

<script>
module.exports = {
  name: 'reasoning',
  computed: {
    nw: function () {
      return nw;
    }
  }
};
</script>
