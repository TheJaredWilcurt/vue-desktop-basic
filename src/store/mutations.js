
const mutations = {
  changeGreeting: function (state, newGreeting) {
    state.greeting = newGreeting;
  }
};

module.exports = mutations;
