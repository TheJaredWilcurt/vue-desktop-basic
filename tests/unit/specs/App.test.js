import Vuex, { Store } from 'vuex';
import VueRouter from 'vue-router';
import { mount, shallowMount, createLocalVue } from '@vue/test-utils';

import app from '@/App';
import features from '@/components/features';
import reasoning from '@/components/reasoning';
import mutations from '@/store/mutations';

describe('App.vue', () => {
  let store;
  let localVue;
  let router;

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.use(Vuex);
    localVue.use(VueRouter);

    store = new Store({
      state: {
        greeting: 'Hi'
      },
      mutations
    });

    router = new VueRouter({
      routes: [
        {
          path: '*',
          redirect: 'reasoning'
        },
        {
          path: '/features',
          name: 'features',
          component: features
        },
        {
          path: '/reasoning',
          name: 'reasoning',
          component: reasoning
        }
      ]
    });
  });

  describe('Creation', () => {
    test('Render correct contents', () => {
      const wrapper = mount(app, {
        store,
        localVue,
        router
      });

      expect(wrapper.find({ ref: 'testApp' }).html())
        .toMatchSnapshot();
    });
  });

  describe('Vuex store', () => {
    test('Modifying the greeting', () => {
      const wrapper = shallowMount(app, {
        store,
        localVue,
        router
      });

      wrapper.vm.$store.commit('changeGreeting', 'Test');

      expect(wrapper.find({ ref: 'testGreeting' }).text())
        .toEqual('Test World');
    });
  });

  describe('Interactions', () => {
    test('Open Dev Tools', () => {
      const wrapper = shallowMount(app, {
        store,
        localVue,
        router
      });

      global.nw = {
        Window: {
          get: jest.fn(() => {
            return {
              showDevTools: jest.fn()
            };
          })
        }
      };

      let devToolsLink = wrapper.find({ ref: 'testShowDevTools' });
      devToolsLink.trigger('click');

      expect(nw.Window.get())
        .toHaveBeenCalled();
    });
  });
});
