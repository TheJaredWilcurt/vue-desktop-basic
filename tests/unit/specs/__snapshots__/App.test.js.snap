// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue Creation Render correct contents 1`] = `
<div id="app">
  <h1>Hi World</h1>
  <nav>
    <!---->
    <a href="#/features" class="">
      features
    </a>
    <a href="#/reasoning" class="router-link-exact-active router-link-active">
      reasoning
    </a>
    |
    <a href="#">Open Dev Tools</a>
  </nav>
  <section>
    <h2>Reasoning:</h2>
    <p><em>Why not use a bundler, like WebPack/Browserify/Parcel/StealJS/Fuse-Box/Etc?</em></p>
    <ul>
      <li>Because bundlers are designed around the limitations of the web. We are working with a different medium, Cross-Platform Desktop Apps (XPDAs).</li>
      <li>The purpose of a bundler is to bundle our code. Features like concatenating, tree-shaking, uglifying, and minifying our code doesn't really offer any benefit. Nor does reducing the number of simultaneous network requests for assets. All assets are
        local, because desktop apps should follow the best practice of "offline-first".</li>
      <li>Ultimately you are shipping an 80+ MB app to your end user, so the tiny optimizations made for the medium of web are quite arbitrary here.</li>
      <li>A much greater concern for desktop apps would be the amount of dependencies in your node_modules folder, since all dependencies, and their dependencies are shipped to the end user (this excludes devDependencies).</li>
      <li>If you wish to use additional features for development (linting, HMR, Sass, etc), I would recommend adding those features via a task runner, like Gulp, or just via simple npm scripts.</li>
    </ul>
    <p><em>Why not use a CDN for Vue/Vue-Router/Vuex/Etc instead of adding them as dependencies?</em></p>
    <ul>
      <li>Because for desktop apps you should always be designing with an "offline-first" mentality. Your app should still run if the user is in Airplane mode.</li>
      <li>Keeping them as dependencies allows you to easily handle version management.</li>
      <li>An alternative option would be to create a script that downloads the desired versions of these libraries into an untracked folder. This would mean you would have the desired versions, and no additional files or dependencies. Then run that script
        as a <code>postinstall</code> npm script. The downside to this is that it would be less familiar to other developers. Though technically more efficient, it is also slightly less maintable and requires documentation for future maintainers/contributors.</li>
    </ul>
  </section>
</div>
`;
